### PR TITLE
Error message if upload is not possible

### LINKED ISSUE #62: Make error message from compiling w/o board selected more user friendly
## Describe the current behavior

On the first use, no board is selected. A new user may not realize it is necessary to select one. If they attempt to compile a sketch without selecting one, they will receive only this cryptic error message:

```
Compilation error: Missing FQBN (Fully Qualified Board Name)
```

and this notification:

```
Error: Request compile failed with message: Compilation error: Missing FQBN (Fully Qualified Board Name)
```

This doesn't mean anything to even very advanced IDE users, and the user will have difficulty understanding what they need to do to resolve it.

## To reproduce

1. Quit the IDE if it is running.
1. Simulate a fresh install by deleting the configuration folder:
   - Windows:
     ```
     C:\Users\<user name>\AppData\Roaming\arduino-ide\
     ```
   - Linux:
     ```
     ~/.config/arduino-ide/
     ```
   - macOS:
     ```
     ~/Library/Application Support/arduino-ide/
     ```
   This step is done because the IDE currently makes it otherwise difficult for those with internal serial ports to achieve the required state of having no port selected, due to the odd fallback port selection behavior (e.g., https://github.com/arduino/arduino-ide/issues/180).
1. Start the Arduino IDE.
   The board/port selector menu should show "**:x: No board selected**".
1. Select **Sketch > Verify/Compile** from the Arduino IDE menus.

## Describe the request

Show a user friendly error message. Something like:

> No board selected. Please select your Arduino board from the **Tools > Board** menu.

## Desktop

- OS: Windows 10
- Version: 2.0.0-rc3-snapshot.1d88263
Date: 2022-01-24T16:22:30.529Z
CLI Version: 0.20.2 [13783819]

## Additional context

Even among Arduino CLI users, I have found that the concept of "FQBN" is not widely known. IDE users are even less likely to be familiar with this acronym, and even less anxious to spend their time learning what it means, since the it is not otherwise a part of the IDE user experience.

---

Note that the error message produced by the reproduction instructions above is currently different than described due to a recent regression in Arduino CLI: https://github.com/arduino/arduino-cli/issues/1812

### Related

- https://github.com/arduino/docs-content/issues/300

### LINKED ISSUE #845: Communicate to user when upload is impossible due to no board selection
## Describe the current behavior

On the first run of a fresh installation, no board is selected in the IDE. A "no board selected" condition might be produced in other ways (e.g., https://github.com/arduino/arduino-ide/issues/657) after that time as well.

Compilation and upload commands are configured for each board. For this reason, compilation or upload is impossible if a board has not been selected.

🙁 The Arduino IDE does not provide any indication of what went wrong when an upload is attempted without a board selected.

## To reproduce

1. Quit the IDE if it is running.
1. Rename or delete (:warning: cautiously) the following folder to simulate a first run:
   - Windows:
     ```
     C:\Users\<user name>\AppData\Roaming\arduino-ide
     ```
   - Linux:
     ```
     ~/.config/arduino-ide
     ```
   - macOS:
     ```
     ~/Library/Application Support/arduino-ide
     ```
1. Start the Arduino IDE.
1. Select a port from the **Tools > Port** menu in the Arduino IDE.
   🛈 This step is not required. It is only done to show that the issue is not dependent on a missing port selection (https://github.com/arduino/arduino-ide/issues/770)
1. Select **Sketch > Upload** from the Arduino IDE menus.

🙁 Nothing at all happens. There is no notification, error message, etc.

## Describe the request

Clearly communicate to the user when uploading is not possible due to no board selection.

Even though it is not great (https://github.com/arduino/arduino-ide/issues/62), there is at least some communication when compilation is attempted without a board selection.

## Desktop

- OS: Windows 10
- Version: 2.0.0-rc4-snapshot.9b1f15d-snapshot-9b1f15d
Date: 2022-02-17T10:40:50.003Z
CLI Version: 0.21.0 [10107d24]

## Additional context

This is likely only another symptom of the same issue that causes https://github.com/arduino/arduino-ide/issues/770

I decided to create a separate issue report since there is a significantly different expected outcome between the two user paths (in https://github.com/arduino/arduino-ide/issues/770 the upload should be successful, while here the issue is only that the failure is not communicated).

---

### Related

- https://github.com/arduino/arduino-ide/issues/973

### PR DESCRIPTION
**Pull request by francescospissu**
### Motivation
If the user presses the upload button and no board is selected, a message should appear warning what went wrong while uploading. Since verify takes place before upload, the same message that is given when the user tries to compile a sketch without selecting a board should be shown. On the other hand, this message is not very clear, so it should be updated.

### Change description
Communicate to user when upload is impossible due to no board selection.
Make error message from compiling without board selected more user friendly.

### Other information
Closes #845.
Closes #62.

This fix depends on the changes made here: https://github.com/arduino/arduino-cli/pull/1848. Without it it is not possible to verify that #62 is resolved.

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Review by kittaakos — APPROVED** [review]
The code looks good to me.

I am not familiar with the board user fields logic, maybe we should change

```ts
if (this.boardRequiresUserFields && !this.cachedUserFields.has(key)) {
```

to 

```ts
```

**Review by ubidefeo — APPROVED** [review]
Works as expected

**Review by davegarthsimpson — APPROVED** [review]

File: arduino-ide-extension/src/browser/contributions/contribution.ts [hunk 1]
```diff
@@ -222,6 +222,10 @@ export abstract class CoreServiceContribution extends SketchContribution {
       } catch {}
     }
     if (message) {
+      if (message.includes('Missing FQBN (Fully Qualified Board Name)')) {
+        message =
+          'No board selected. Please select your Arduino board from the Tools > Board menu.';
```
Comment (per1234) [inline arduino-ide-extension/src/browser/contributions/contribution.ts:227]:
```suggestion
        message = nls.localize(
          'arduino/coreContribution/noBoardSelected',
          'No board selected. Please select your Arduino board from the Tools > Board menu.'
        );
```

**Review by per1234 — CHANGES_REQUESTED** [review]

File: arduino-ide-extension/src/browser/contributions/contribution.ts (hunk 1 above)
Comment (francescospissu) [reply arduino-ide-extension/src/browser/contributions/contribution.ts:227]: Thanks for your suggestion @per1234!

**Review by per1234 — APPROVED** [review]
Thanks Francesco!

**Pull request merged**